drivers/dma/dma_dw_axi.c

```diff
@@ -0,0 +1,1002 @@
+
+/* source burst transaction length */
+#define DMA_DW_AXI_CTL_SRC_MSIZE(x)              FIELD_PREP(GENMASK64(17, 14), x)
+/* destination burst transaction length */
+#define DMA_DW_AXI_CTL_DST_MSIZE(x)              FIELD_PREP(GENMASK64(21, 18), x)
+/* source transfer width */
+#define DMA_DW_AXI_CTL_SRC_WIDTH(x)              FIELD_PREP(GENMASK64(10, 8), x)
+/* destination transfer width */
+#define DMA_DW_AXI_CTL_DST_WIDTH(x)              FIELD_PREP(GENMASK64(13, 11), x)
+
+/* mask all the interrupts */
+#define DMA_DW_AXI_IRQ_NONE                      0
+/* enable block completion transfer interrupt */
+#define DMA_DW_AXI_IRQ_BLOCK_TFR                 BIT64(0)
+/* enable transfer completion interrupt */
+#define DMA_DW_AXI_IRQ_DMA_TFR                   BIT64(1)
+/* enable interrupts on any dma transfer error */
+#define DMA_DW_AXI_IRQ_ALL_ERR                   (GENMASK64(14, 5) | GENMASK64(21, 16))
+
+/* global enable bit for dma controller */
+#define DMA_DW_AXI_CFG_EN                        BIT64(0)
+/* global enable bit for interrupt */
+#define DMA_DW_AXI_CFG_INT_EN                    BIT64(1)
+
+/* descriptor used by dw axi dma controller*/
+struct dma_lli {
+	uint64_t sar;
+	uint64_t dar;
+	uint32_t block_ts_lo;
+	uint32_t reserved;
+	uint64_t llp;
+	uint64_t ctl;
+	uint32_t sstat;
+	uint32_t dstat;
+	uint64_t llp_status;
+	uint64_t reserved1;
+} __aligned(64);
+
+/* status of the channel */
+enum dma_dw_axi_channel_state {
+	DMA_DW_AXI_CH_IDLE,
+	DMA_DW_AXI_CH_SUSPENDED,
+	DMA_DW_AXI_CH_ACTIVE,
+	DMA_DW_AXI_CH_PREPARED,
+};
+
+/* source and destination transfer width */
+enum dma_dw_axi_ch_width {
+	BITS_8,
+	BITS_16,
+	BITS_32,
+	BITS_64,
+	BITS_128,
+	BITS_256,
+	BITS_512,
+};
+
+/* transfer direction and flow controller */
+enum dma_dw_axi_tt_fc {
+	M2M_DMAC,
+	M2P_DMAC,
+	P2M_DMAC,
+	P2P_DMAC,
+	P2M_SRC,
+	P2P_SRC,
+	M2P_DST,
+	P2P_DST,
+};
+
+/* type of multi-block transfer */
+enum dma_dw_axi_multi_blk_type {
+	MULTI_BLK_CONTIGUOUS,
+	MULTI_BLK_RELOAD,
+	MULTI_BLK_SHADOW_REG,
+	MULTI_BLK_LLI,
+};
+
+/* dma driver channel specific information */
+struct dma_dw_axi_ch_data {
+	/* lli descriptor base */
+	struct dma_lli *lli_desc_base;
+	/* lli current descriptor */
+	struct dma_lli *lli_desc_current;
+	/* dma channel state */
+	atomic_t ch_state;
+	/* direction of transfer */
+	uint32_t direction;
+	/* number of descriptors */
+	uint32_t lli_desc_count;
+	/* cfg register configuration for dma transfer */
+	uint64_t cfg;
+	/* mask and unmask interrupts */
+	uint64_t irq_unmask;
+	/* user call back for dma transfer completion */
+	dma_callback_t dma_xfer_callback;
+	/* user data for dma callback for dma transfer completion */
+	void *priv_data_xfer;
+	/* user call back for dma block transfer completion */
+	dma_callback_t dma_blk_xfer_callback;
+	/* user data for dma callback for dma block transfer completion */
+	void *priv_data_blk_tfr;
+	/* mutex for chan data */
+	struct k_mutex ch_mutex_lock;
```

Comment:
intel_socfpga_agilex5_socdk has 2 instance of dma controller with 4 channels enabled per instance. 
And there can be 48 possible number of users (may not be all active at a time but possible to have more than 8 users at a time depending on use cases)

This is the reason we have implemented mutual exclusion to avoid any synchronization issues